File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/proc/AttProcService.java
package ru.citeck.ecos.records3.record.op.atts.service.proc;

import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.data.DataValue;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.records2.RecordsServiceFactory;
import ru.citeck.ecos.records3.record.op.atts.service.schema.SchemaAtt;
import ru.citeck.ecos.records3.record.op.atts.service.schema.read.AttSchemaReader;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
public class AttProcService {

    private final static String PROC_ATT_ALIAS_PREFIX = "__proc_att_";

    private final Map<String, AttProcessor> processors = new ConcurrentHashMap<>();
    private final AttSchemaReader attSchemaReader;

    public AttProcService(RecordsServiceFactory serviceFactory) {
        attSchemaReader = serviceFactory.getAttSchemaReader();
    }

    public DataValue process(ObjectData meta,
                             DataValue value,
                             List<AttProcDef> processorsDef) {


        if (processorsDef.size() == 0) {
            return value;
        }

        DataValue newValue = value;

        for (AttProcDef def : processorsDef) {
            AttProcessor proc = getProcessor(def.getType());
            if (proc != null) {
                Object newValueObj = proc.process(meta, newValue, def.getArguments());
                if (newValueObj instanceof String) {
                    newValue = DataValue.createStr(newValueObj);
                } else {
                    newValue = DataValue.create(newValueObj);
                }
            }
        }

        if (!Objects.equals(value, newValue)) {
            value = newValue;
        }

        return value;
    }

    private Set<String> getAttsToLoadSet(List<AttProcDef> processorsDef) {

        if (processorsDef.isEmpty()) {
            return Collections.emptySet();
        }

        Set<String> attributes = new HashSet<>();

        for (AttProcDef def : processorsDef) {
            AttProcessor proc = getProcessor(def.getType());
            if (proc != null) {
                attributes.addAll(proc.getAttributesToLoad(def.getArguments()));
            }
        }

        return attributes;
    }

    @NotNull
    public List<SchemaAtt> getProcessorsAtts(List<SchemaAtt> attributes) {

        Set<String> attributesToLoad = new HashSet<>();

        for (SchemaAtt att : attributes) {
            attributesToLoad.addAll(getAttsToLoadSet(att.getProcessors()));
        }

        if (attributesToLoad.isEmpty()) {
            return Collections.emptyList();
        }

        Map<String, String> procAtts = new HashMap<>();

        for (String att : attributesToLoad) {
            procAtts.put(PROC_ATT_ALIAS_PREFIX + att, att);
        }

        return attSchemaReader.read(procAtts);
    }

    public ObjectData applyProcessors(ObjectData data, Map<String, List<AttProcDef>> processors) {

        if (processors.isEmpty()) {
            return data;
        }

        Map<String, Object> dataMap = new HashMap<>();
        data.forEach(dataMap::put);

        return ObjectData.create(applyProcessors(dataMap, processors));
    }

    public Map<String, Object> applyProcessors(Map<String, Object> data, Map<String, List<AttProcDef>> processors) {

        ObjectData procData = ObjectData.create();
        Map<String, Object> resultData = new HashMap<>();

        data.forEach((k, v) -> {
            if (k.startsWith(PROC_ATT_ALIAS_PREFIX)) {
                procData.set(k.replaceFirst(PROC_ATT_ALIAS_PREFIX, ""), v);
            } else {
                resultData.put(k, v);
            }
        });

        if (processors.isEmpty()) {
            return resultData;
        }

        processors.forEach((att, attProcessors) -> {
            if (!attProcessors.isEmpty()) {
                DataValue value = DataValue.create(resultData.get(att));
                resultData.put(att, process(procData, value, attProcessors));
            }
        });

        return resultData;
    }

    @Nullable
    private AttProcessor getProcessor(String type) {
        AttProcessor proc = processors.get(type);
        if (proc == null) {
            log.error("Attribute processor doesn't found for type '" + type + "'");
        }
        return proc;
    }

    public void register(AttProcessor processor) {
        this.processors.put(processor.getType(), processor);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/proc/AttProcService.kt
package ru.citeck.ecos.records3.record.atts.proc

import mu.KotlinLogging
import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.atts.schema.SchemaAtt
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.LinkedHashMap

class AttProcService(serviceFactory: RecordsServiceFactory) {

    companion object {
        private const val PROC_ATT_ALIAS_PREFIX: String = "__proc_att_"

        private val log = KotlinLogging.logger {}
    }

    private val processors = ConcurrentHashMap<String, AttProcessor>()
    private val attSchemaReader = serviceFactory.attSchemaReader

    fun process(attributes: ObjectData, value: DataValue, processorsDef: List<AttProcDef>): DataValue? {

        if (processorsDef.isEmpty()) {
            return value
        }
        var newValue: DataValue = value
        for (def in processorsDef) {
            val proc: AttProcessor? = getProcessor(def.type)
            if (proc != null) {
                val newValueObj: Any? = proc.process(attributes, newValue, def.arguments)
                newValue = if (newValueObj is String) {
                    DataValue.createStr(newValueObj)
                } else {
                    DataValue.create(newValueObj)
                }
            }
        }
        return if (value != newValue) {
            newValue
        } else {
            value
        }
    }

    private fun getAttsToLoadSet(processorsDef: List<AttProcDef>): Set<String> {

        if (processorsDef.isEmpty()) {
            return emptySet()
        }
        val attributes: MutableSet<String> = HashSet()
        for (def in processorsDef) {
            val proc: AttProcessor? = getProcessor(def.type)
            if (proc != null) {
                attributes.addAll(proc.getAttsToLoad(def.arguments))
            }
        }
        return attributes
    }

    fun getProcessorsAtts(attributes: List<SchemaAtt>): List<SchemaAtt> {

        val attributesToLoad: MutableSet<String> = HashSet()
        for (att in attributes) {
            attributesToLoad.addAll(getAttsToLoadSet(att.processors))
        }
        if (attributesToLoad.isEmpty()) {
            return emptyList()
        }
        val procAtts: MutableMap<String, String> = LinkedHashMap()
        for (att in attributesToLoad) {
            procAtts[PROC_ATT_ALIAS_PREFIX + att] = att
        }
        return attSchemaReader.read(procAtts)
    }

    fun applyProcessors(data: ObjectData, processors: Map<String, List<AttProcDef>>): ObjectData {

        if (processors.isEmpty()) {
            return data
        }
        val dataMap: MutableMap<String, Any?> = LinkedHashMap()
        data.forEach {
            key, value ->
            dataMap[key] = value
        }
        return ObjectData.create(applyProcessors(dataMap, processors))
    }

    fun applyProcessors(data: Map<String, Any?>, processors: Map<String, List<AttProcDef>>): Map<String, Any?> {

        val procData: ObjectData = ObjectData.create()
        val resultData: MutableMap<String, Any?> = LinkedHashMap()

        data.forEach { (k, v) ->
            if (k.startsWith(PROC_ATT_ALIAS_PREFIX)) {
                procData.set(k.replaceFirst(PROC_ATT_ALIAS_PREFIX.toRegex(), ""), v)
            } else {
                resultData[k] = v
            }
        }
        if (processors.isEmpty()) {
            return resultData
        }

        processors.forEach { (att, attProcessors) ->
            if (attProcessors.isNotEmpty()) {
                val value: DataValue = DataValue.create(resultData[att])
                resultData[att] = process(procData, value, attProcessors)
            }
        }
        return resultData
    }

    private fun getProcessor(type: String): AttProcessor? {
        val proc: AttProcessor? = processors[type]
        if (proc == null) {
            log.error("Attribute processor doesn't found for type '$type'")
        }
        return proc
    }

    fun register(processor: AttProcessor) {
        processors[processor.getType()] = processor
    }
}
